### ISSUE TITLE
该工程的使用说明

### ISSUE THREAD

**Issue by Vanillaholic** [paulchen2713/DL_CFAR#1]
您好，抱歉打扰您，想问一下有没有该工程以及数据集生成的使用说明？期待您的回复

**Comment by paulchen2713** [comment]
您好，因為東西年久失修、很多細節我也不記得了，如果您需要的是可以馬上運作的 solution 我可能幫不上忙，但概念上就是用合成雷達資料 (synthetic RD Map data) 去訓練神經網路分類器 (classifier)，分類有無或有幾個雷達目標，網路架構可以從 .py 檔 source code 看到，之前 (5-6 年前) 應該是用 tensorflow 1.x 或 keras 1.0 (原生 keras) 做的，現在很多 API 已經改版或被棄用。

如果您想要再次重現的話，新刻的神經網路接口要特別處理，要能直接吃 合成雷達資料 / 矩陣 / np array / matlab matrix，可以想像成輸入資料是 值域不固定的灰階影像，如果直接從一般的圖片分類器改過來需要注意維度問題，因為這些網路一般是先讀 RGB 進來再轉 np array 然後每個 pixel 有 3 維，但實際的 RD Map 合成資料每個 pixel 只有 1 維。

另一個 repo - DL_CFAR_data 裡面有模擬 RD Map 的 Matlab code (可能也沒辦法直接用)，模擬方式是隨機設定 Range 跟 Doppler (velocity) 跟 目標點數，然後直接做 2D FFT 模擬生成 Range-Doppler Map，再看要存成 .mat 陣列、灰階或彩色 .jpg 的訓練資料，標籤解答也是已知，因為在產資料的時候我們已經知道這張 RD Map 有無目標或有幾個目標。

此外，這個模擬沒有管時序性，也就是 frame 跟 frame 之間可以不連續 (因為原先目的只是為了訓練分類器)，所以很不實際，水一篇文章還可以，但在解決實際工程問題上~~可能~~沒什麼用。

希望這些資訊有提供您一點幫助。

**Comment by Vanillaholic** [comment]
@paulchen2713 謝謝您的回覆，但是這篇論文有個地方不是很懂，文章中分析了不同False Alarm Rate下的檢測率，但是False Alarm Rate是如何確定的？請問您知道嗎？

**Comment by paulchen2713** [comment]
您好，False Alarm Rate 其實就只是一個比值: (detector 或 model 說有目標，實際上沒目標的數量) / (所有有目標的數量)，然後這是一個相對的概念，False Alarm Rate 的優劣是相同情況不同算法比較才有意義，也就是 False Alarm Rate 多低才夠低，是您要去論證說明，或是採某個領域大家的共識

撇除這些，具體上有 2 個方案，一是以單張 RD map 為單位；二是以單個 pixel 為單位

所以方案一就是，假設你的測試集 (test set) 有 2 萬張 RD Map (假設都是有目標) 及每張的 label (有無目標 或 幾個目標)，當你跑完測試之後就可以算得 "detector 偵測錯的張數" 除以 "所有有目標的張數 (也就是總張數)" 即 False Alarm Rate，比如有 2 張 detector 說沒目標，那 False Alarm Rate = 2/20000 = 0.1%

當然你的測試集也可以混合有目標跟無目標的 RD Map (混合方案不推薦) 或是全部都是 "沒有目標"，這樣 False Alarm Rate 就是 "detector 說有目標的張數" 除以 "總張數" 

但讀到這邊就會發現一個問題，因為通常在通訊或雷達系統下，有意義的 False Alarm Rate 的量級通常是 10^-5 或更小 (然後 SNR 是某個 dB)，這樣一來即使先不管 SNR 是多少，你在本地端得存幾十萬或幾百萬張 RD map，通常 1 個 pixel 至少是 1 byte (取決於檔案格式)，假設你的 RD map 才 10x10，一張 100 byte，一萬張就要 1 MB

這時候就可以引入方案二，只需要改變 label 跟 detector 輸出結果，首先 label 需要存一個 bit map 記錄哪個 pixel 是目標哪些不是，通常可以用一個 0/1 bit map 來存；然後 detector 輸出不能像 classifier 只輸出 class label 而是要保存每個 pixel 的偵測結果，也就是輸出一個 0/1 bit map，這樣最少只需要幾千張 RD map 就能統計到 10^-5 或更小的 False Alarm Rate，當然還要取決於 detector 本身的 accuracy，但本質上這樣做可以稀釋 False Alarm Rate 

最後選什麼方案取決於你的目標(當然，還有演算法特性，學理上這兩種方法都不算作弊，理論上是對的，但實務上~不一定~沒用

**Comment by Vanillaholic** [comment]
@paulchen2713 您好，感謝您的回覆，我的意思是 **constant** false alarm rate，是在保持恆定虛警概率下的檢測概率（類似 CA-CFAR、OS-CFAR 算法）。但如果用 CNN 的話，如何設置相關參數以保持 false alarm rate 恆定呢？

**Comment by paulchen2713** [comment]
您好，從文章來看其實沒有一個完整的 false alarm rate 對應 scaling factor 的公式，也不知道怎麼建立 CNN 架構對 **constant** false alarm rate 的關係，也就是沒辦法做到 "假設如果想要達到某個 false alarm rate，那 CNN 應該怎麼設計"

**Comment by Vanillaholic** [comment]
@paulchen2713 所以这篇文章根本不是恒虚警概率是吗？只是预测单个frame噪声的一个CNN？

**Comment by paulchen2713** [comment]
嚴格來講可以算是恆虛警率，因為訓練好的 CNN 在測試的時候，可以達到某個 "恆定" 的虛警率 (給定 desired dection rate 及 SNR 的情況下)。
您說的 "預測單個 frame 的 noise level" 的確是這個 CNN 的行為，CA-CFAR, OS-CFAR 核心思想不也是如何估計 noise level 嗎?
